File: tests/typescript/tests-typescript-playwright/src/pages/login.page.ts
import type { Locator, Page } from '@playwright/test';

export class LoginPage {
  readonly page: Page;
  readonly loginForm: Locator;
  readonly loginInput: Locator;
  readonly passwordInput: Locator;
  readonly submitButton: Locator;
  readonly formTitle: Locator;
  readonly errorMessage: Locator;
  readonly registerLink: Locator;

  constructor(page: Page) {
    this.page = page;
    this.loginForm = page.getByTestId('login-form');
    this.loginInput = page.getByTestId('login-input');
    this.passwordInput = page.getByTestId('password-input');
    this.submitButton = page.getByTestId('submit-button');
    this.formTitle = page.getByTestId('login-form-title');
    this.errorMessage = page.getByTestId('error-message');
    this.registerLink = page.getByTestId('register-link');
  }

  async open(): Promise<void> {
    // Relative to baseURL (ends with '/') — stays inside path-mounted deploys.
    await this.page.goto('login');
    await this.shouldBeOpen();
  }

  async shouldBeOpen(): Promise<void> {
    await this.loginForm.waitFor({ state: 'visible' });
  }

  async shouldShowLoginForm(): Promise<void> {
    await this.formTitle.waitFor({ state: 'visible' });
    await this.loginInput.waitFor({ state: 'visible' });
    await this.passwordInput.waitFor({ state: 'visible' });
    await this.submitButton.waitFor({ state: 'visible' });
  }

  async login(username: string, password: string): Promise<void> {
    await this.loginInput.fill(username);
    await this.passwordInput.fill(password);
    await this.submitButton.click();
  }

  async typeUsername(username: string): Promise<void> {
    await this.loginInput.fill(username);
  }

  async typePassword(password: string): Promise<void> {
    await this.passwordInput.fill(password);
  }

  async submitExpectingError(): Promise<void> {
    await this.submitButton.click();
    await this.errorMessage.waitFor({ state: 'visible' });
  }

  async clickRegisterLink(): Promise<void> {
    await this.registerLink.click();
  }
}

File: tests/go/tests-go-testing-allure3-playwright/login_page.go
package tests

import (
	allure "github.com/allure-framework/allure-go/commons/gotest"
	"github.com/mxschmitt/playwright-go"
	"github.com/stretchr/testify/require"
)

type LoginPage struct {
	app  *App
	page playwright.Page
}

func (p *LoginPage) Form() playwright.Locator {
	return p.page.GetByTestId("login-form")
}

func (p *LoginPage) LoginInput() playwright.Locator {
	return p.page.GetByTestId("login-input")
}

func (p *LoginPage) PasswordInput() playwright.Locator {
	return p.page.GetByTestId("password-input")
}

func (p *LoginPage) Submit() playwright.Locator {
	return p.page.GetByTestId("submit-button")
}

func (p *LoginPage) FormTitle() playwright.Locator {
	return p.page.GetByTestId("login-form-title")
}

func (p *LoginPage) ErrorMessage() playwright.Locator {
	return p.page.GetByTestId("error-message")
}

func (p *LoginPage) RegisterLink() playwright.Locator {
	return p.page.GetByTestId("register-link")
}

func (p *LoginPage) Open() *LoginPage {
	p.app.t.Helper()
	p.app.a.Step("Open login page", func(*allure.Context) {
		_, err := p.page.Goto("login")
		require.NoError(p.app.t, err)
		p.ShouldBeOpen()
	})
	return p
}

func (p *LoginPage) ShouldBeOpen() *LoginPage {
	p.app.t.Helper()
	waitVisible(p.app.t, p.Form())
	return p
}

func (p *LoginPage) ShouldShowLoginForm() *LoginPage {
	p.app.t.Helper()
	waitVisible(p.app.t, p.FormTitle())
	waitVisible(p.app.t, p.LoginInput())
	waitVisible(p.app.t, p.PasswordInput())
	waitVisible(p.app.t, p.Submit())
	return p
}

func (p *LoginPage) Login(username, password string) *LoginPage {
	p.app.t.Helper()
	require.NoError(p.app.t, p.LoginInput().Fill(username))
	require.NoError(p.app.t, p.PasswordInput().Fill(password))
	require.NoError(p.app.t, p.Submit().Click())
	return p
}

func (p *LoginPage) TypeUsername(username string) *LoginPage {
	p.app.t.Helper()
	require.NoError(p.app.t, p.LoginInput().Fill(username))
	return p
}

func (p *LoginPage) TypePassword(password string) *LoginPage {
	p.app.t.Helper()
	require.NoError(p.app.t, p.PasswordInput().Fill(password))
	return p
}

func (p *LoginPage) SubmitExpectingError() *LoginPage {
	p.app.t.Helper()
	require.NoError(p.app.t, p.Submit().Click())
	waitVisible(p.app.t, p.ErrorMessage())
	return p
}

func (p *LoginPage) ClickRegisterLink() *LoginPage {
	p.app.t.Helper()
	require.NoError(p.app.t, p.RegisterLink().Click())
	return p
}

func (p *LoginPage) Reload() *LoginPage {
	p.app.t.Helper()
	_, err := p.page.Reload()
	require.NoError(p.app.t, err)
	return p.ShouldBeOpen()
}
